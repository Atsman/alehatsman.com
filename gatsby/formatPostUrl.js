module.exports = (id) => {
  return `/posts/${id}.html`
}
