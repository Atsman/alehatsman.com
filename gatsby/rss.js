const formatPostUrl = require('./formatPostUrl')

module.exports = {
  resolve: 'gatsby-plugin-feed',
  options: {
    query: `
          {
            site {
              siteMetadata {
                title
                description
                siteUrl
                site_url: siteUrl
              }
            }
          }
        `,
    feeds: [
      {
        serialize: ({ query: { site, allMdx } }) => {
          return allMdx.edges.map((edge) => {
            const imageSrc = edge.node.frontmatter.featuredImage.childImageSharp.fixed.src
            return Object.assign({}, edge.node.frontmatter, {
              description: edge.node.frontmatter.description,
              date: edge.node.frontmatter.createdAt,
              url:
                site.siteMetadata.siteUrl +
                formatPostUrl(edge.node.frontmatter.id),
              guid:
                site.siteMetadata.siteUrl +
                formatPostUrl(edge.node.frontmatter.id),
              custom_elements: [{ 'content:encoded': edge.node.html }],
              enclosure: imageSrc && {
                url: site.siteMetadata.siteUrl + imageSrc
              }
            })
          })
        },
        query: `
          {
            allMdx(filter: {
              frontmatter: {public: {eq: true}}}, 
              sort: {fields: frontmatter___createdAt, order: DESC}
            ) {
              edges {
                node {
                  html
                  frontmatter {
                    id
                    createdAt
                    description
                    public
                    tags
                    title
                    updatedAt
                    featuredImage {
                      childImageSharp {
                        fixed {
                          src
                        }
                      }
                    }
                  }
                }
              }
            }
          }
            `,
        output: '/rss.xml',
        title: 'Aleh Atsman | Software Engineer Blog'
      }
    ]
  }
}
