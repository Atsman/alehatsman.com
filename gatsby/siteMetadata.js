module.exports = {
  author: 'Aleh Atsman',
  description:
  'Computer hacker from young childhood. Full-stack software engineer. Freelancer. Solve problems, deliver high-quality software, write about programming.',
  email: 'aleh.atsman@gmail.com',
  github: 'https://github.com/atsman',
  image: 'images/me2.jpg',
  linkedin: 'https://linkedin.com/in/alehatsman',
  telegram: 'https://t.me/alehatsman',
  title: 'Software Engineer Blog',
  titleTemplate: 'Aleh Atsman | %s',
  twitter: 'https://twitter.com/alehatsman',
  twitterUsername: '@alehatsman',
  url: 'https://alehatsman.com',
  siteUrl: 'https://alehatsman.com'
}
