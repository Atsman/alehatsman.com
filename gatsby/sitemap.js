module.exports = {
  resolve: 'gatsby-plugin-sitemap',
  options: {
    output: '/sitemap.xml',
    query: `
        {
          site {
            siteMetadata {
              siteUrl
            }
          }

          allSitePage {
            edges {
              node {
                path
              }
            }
          }
      }`,
    serialize: ({ site, allSitePage }) =>
      allSitePage.edges.map(edge => {
        return {
          url: site.siteMetadata.siteUrl + edge.node.path,
          changefreq: 'daily',
        }
      })
  }
}
