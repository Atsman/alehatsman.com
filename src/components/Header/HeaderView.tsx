import React, { FC } from 'react'
import { FaGithubAlt, FaAt, FaLinkedin } from 'react-icons/fa'
import css from '@styled-system/css'

import { Link } from '@/components/Link'
import { Box } from '@/components/Box'
import { Text } from '@/components/Text'
import { Spacer } from '@/components/Spacer'

export interface Props {
  data: {
    site: {
      siteMetadata: {
        author: string
        email: string
        github: string
        linkedin: string
      }
    }
  }
}

const formatLabel = (url: string) => {
  const { host, pathname } = new URL(url)
  return `${host}${pathname}`
}

interface LinkProps {
  Icon: any
  href: string
  label: string
}

const ExternalLink: FC<LinkProps> = ({ Icon, href, label, ...props }) => (
  <Box
    display="flex"
    alignItems="center"
    color="gray.2"
    {...props}
  >
    <Link
      href={href}
      css={css({
        display: 'flex',
        alignItems: 'center',
        py: [2, 0],
        fontSize: 1
      })}
    >
      <Icon />
      <Spacer ml={1} />
      {label}
    </Link>
  </Box>
)

export const HeaderView: FC<Props> = ({ data }) => {
  const {
    author,
    email,
    github,
    linkedin
  } = data.site.siteMetadata

  return (
    <Box
      display="flex"
      flexDirection={['column', 'row']}
      justifyContent="space-between"
    >
      <Text
        fontSize={2}
        lineHeight="copy"
        fontWeight="bold"
        color="black"
      >
        {author}
      </Text>
      <Box
        display="flex"
        flexDirection={['column', 'row']}
      >
        <ExternalLink
          Icon={FaGithubAlt}
          href={github}
          label={formatLabel(github)}
        />
        <Spacer ml={[0, 3]} />
        <ExternalLink
          Icon={FaAt}
          href={`mailto:${email}`}
          label={email}
        />
        <Spacer ml={[0, 3]} />
        <ExternalLink
          Icon={FaLinkedin}
          href={linkedin}
          label={formatLabel(linkedin)}
        />
      </Box>
    </Box>
  )
}
