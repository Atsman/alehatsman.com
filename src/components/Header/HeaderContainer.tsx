import React, { FC, ComponentType, ComponentProps } from 'react'
import { useStaticQuery, graphql } from 'gatsby'
import { HeaderView } from './HeaderView'

type HeaderProps = ComponentProps<typeof HeaderView>

interface Props {
  Presenter: ComponentType<HeaderProps>
}

export const query = graphql`
  query {
    site {
      siteMetadata {
        author
        email
        github
        image
        linkedin
        twitter
      }
    }
  }
`

const dataToHeaderProps = (data): HeaderProps => {
  return {
    ...data.site.siteMetadata,
  }
}

export const HeaderContainer: FC<Props> = ({ Presenter }) => {
  const data = useStaticQuery(query)
  return <Presenter {...dataToHeaderProps(data)} />
}
