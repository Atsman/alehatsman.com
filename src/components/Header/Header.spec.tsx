import React from 'react'
import { render } from '@testing-library/react'
import { useStaticQuery } from 'gatsby'
import { HeaderView } from './HeaderView'
import { HeaderContainer } from './HeaderContainer'

const data = {
  site: {
    siteMetadata: {
      author: 'Aleh Atsman',
      email: 'aleh.atsman@gmail.com',
      github: 'https://github.com/atsman',
      linkedin: 'https://linkedin.com/in/alehatsman'
    }
  }
}

describe('HeaderContainer', () => {
  it('should pass data to Presenter', () => {
    (useStaticQuery as jest.MockedFunction<typeof useStaticQuery>).mockReturnValue(data)
    const presenter = jest.fn(() => (<div></div>))
    render(<HeaderContainer Presenter={presenter} />)
    expect(presenter).toBeCalledWith({ data }, {})
  })
})

describe('HeaderView', () => {
  it('should match snapshot', () => {
    expect(render(<HeaderView data={data} />).asFragment()).toMatchSnapshot()
  })

  it('should render all information', async () => {
    const container = render(<HeaderView data={data} />)

    const author = await container.findByText('Aleh Atsman')
    const github = await container.findByText('github.com/atsman')
    const email = await container.findByText('aleh.atsman@gmail.com')
    const linkedin = await container.findByText('linkedin.com/in/alehatsman')

    expect(author).toBeInTheDocument()
    expect(github).toHaveAttribute('href', 'https://github.com/atsman')
    expect(email).toHaveAttribute('href', 'mailto:aleh.atsman@gmail.com')
    expect(linkedin).toHaveAttribute(
      'href',
      'https://linkedin.com/in/alehatsman'
    )
  })
})
