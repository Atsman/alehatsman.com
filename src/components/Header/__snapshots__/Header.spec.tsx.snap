// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderView should match snapshot 1`] = `
<DocumentFragment>
  .emotion-26 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media screen and (min-width:40em) {
  .emotion-26 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

.emotion-0 {
  font-size: 16px;
  line-height: copy;
  font-weight: bold;
}

.emotion-24 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media screen and (min-width:40em) {
  .emotion-24 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

.emotion-6 {
  box-sizing: border-box;
  min-width: 0;
  color: gray.2;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 8px;
  padding-bottom: 8px;
  font-size: 14px;
}

@media screen and (min-width:40em) {
  .emotion-4 {
    padding-top: 0;
    padding-bottom: 0;
  }
}

.emotion-2 {
  margin-left: 4px;
}

.emotion-8 {
  margin-left: 0;
}

@media screen and (min-width:40em) {
  .emotion-8 {
    margin-left: 16px;
  }
}

<div
    class="emotion-26 emotion-7"
    display="flex"
  >
    <span
      class="emotion-0 emotion-1"
      color="black"
      font-size="2"
      font-weight="bold"
    >
      Aleh Atsman
    </span>
    <div
      class="emotion-24 emotion-7"
      display="flex"
    >
      <div
        class="emotion-6 emotion-7"
        color="gray.2"
        display="flex"
      >
        <a
          class="emotion-4 emotion-5"
          href="https://github.com/atsman"
          target="_blank"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 480 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
            />
          </svg>
          <div
            class="emotion-2 emotion-3"
          />
          github.com/atsman
        </a>
      </div>
      <div
        class="emotion-8 emotion-3"
      />
      <div
        class="emotion-6 emotion-7"
        color="gray.2"
        display="flex"
      >
        <a
          class="emotion-4 emotion-5"
          href="mailto:aleh.atsman@gmail.com"
          target="_blank"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C118.941 8 8 118.919 8 256c0 137.059 110.919 248 248 248 48.154 0 95.342-14.14 135.408-40.223 12.005-7.815 14.625-24.288 5.552-35.372l-10.177-12.433c-7.671-9.371-21.179-11.667-31.373-5.129C325.92 429.757 291.314 440 256 440c-101.458 0-184-82.542-184-184S154.542 72 256 72c100.139 0 184 57.619 184 160 0 38.786-21.093 79.742-58.17 83.693-17.349-.454-16.91-12.857-13.476-30.024l23.433-121.11C394.653 149.75 383.308 136 368.225 136h-44.981a13.518 13.518 0 0 0-13.432 11.993l-.01.092c-14.697-17.901-40.448-21.775-59.971-21.775-74.58 0-137.831 62.234-137.831 151.46 0 65.303 36.785 105.87 96 105.87 26.984 0 57.369-15.637 74.991-38.333 9.522 34.104 40.613 34.103 70.71 34.103C462.609 379.41 504 307.798 504 232 504 95.653 394.023 8 256 8zm-21.68 304.43c-22.249 0-36.07-15.623-36.07-40.771 0-44.993 30.779-72.729 58.63-72.729 22.292 0 35.601 15.241 35.601 40.77 0 45.061-33.875 72.73-58.161 72.73z"
            />
          </svg>
          <div
            class="emotion-2 emotion-3"
          />
          aleh.atsman@gmail.com
        </a>
      </div>
      <div
        class="emotion-8 emotion-3"
      />
      <div
        class="emotion-6 emotion-7"
        color="gray.2"
        display="flex"
      >
        <a
          class="emotion-4 emotion-5"
          href="https://linkedin.com/in/alehatsman"
          target="_blank"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
            />
          </svg>
          <div
            class="emotion-2 emotion-3"
          />
          linkedin.com/in/alehatsman
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
